commit 3b3f510111b1e7ac36b8db6c66f729abdbf4fd87 (refs/changes/30/1130/8)
Author: Orlando Vazquez <ovazquez@gmail.com>
Date:   2016-12-15T15:36:03-08:00 (2 years, 10 months ago)
    
    AGENT-1050 docker exec interactive terminal hangs after a while
    Reviewed by: Todd Whiteman <todd.whiteman@joyent.com>
    Reviewed by: Josh Wilsdon <josh@wilsdon.ca>
    Approved by: Todd Whiteman <todd.whiteman@joyent.com>

diff --git a/bin/docker-build.js b/bin/docker-build.js
index cdacea1..8071b6d 100644
--- a/bin/docker-build.js
+++ b/bin/docker-build.js
@@ -30,7 +30,7 @@ var rimraf = require('rimraf');
 var sprintf = require('sprintf').sprintf;
 var zfs = require('zfs').zfs;
 
-var LineStream = require('../lib/linestream');
+var LineStream = require('lstream');
 var smartDcConfig = require('../lib/task_agent/smartdc-config');
 
 
@@ -330,31 +330,37 @@ function buildFromContext(opts, callback) {
     // to a request the build system has made.
     var buildEventStream = new LineStream();
     socket.pipe(buildEventStream);
-    buildEventStream.on('line', function (event) {
-        log.debug('client event received: %j', event);
-        var id;
-        var cbEvent;
-        try {
-            event = JSON.parse(event);
-        } catch (e) {
-            log.error('Build: invalid json: %s - ignoring', event);
-            return;
-        }
-        switch (event.type) {
-            case 'callback':
-                id = event.messageId;
-                cbEvent = pendingCallbackEvents[id];
-                assert.object(cbEvent, 'cbEvent with messageId ' + id);
-                delete pendingCallbackEvents[id];
-                if (event.error) {
-                    cbEvent.callback(new Error(event.error));
-                } else {
-                    cbEvent.callback(null, event.result);
-                }
-                break;
-            default:
-                log.error('Unhandled socket event - ignoring: %j', event);
-                break;
+
+    buildEventStream.on('readable', function buildFromContextOnReadable() {
+        var line;
+
+        while ((line = buildEventStream.read()) != null) {
+            log.debug('client event received: %j', line);
+            var id;
+            var cbEvent;
+            var event;
+            try {
+                event = JSON.parse(line);
+            } catch (e) {
+                log.error('Build: invalid json: %s - ignoring', line);
+                continue;
+            }
+            switch (event.type) {
+                case 'callback':
+                    id = event.messageId;
+                    cbEvent = pendingCallbackEvents[id];
+                    assert.object(cbEvent, 'cbEvent with messageId ' + id);
+                    delete pendingCallbackEvents[id];
+                    if (event.error) {
+                        cbEvent.callback(new Error(event.error));
+                    } else {
+                        cbEvent.callback(null, event.result);
+                    }
+                    break;
+                default:
+                    log.error('Unhandled socket event - ignoring: %j', event);
+                    break;
+            }
         }
     });
 
@@ -544,31 +550,37 @@ function commitImage(opts, callback) {
     // to a request the build system has made.
     var commitEventStream = new LineStream();
     socket.pipe(commitEventStream);
-    commitEventStream.on('line', function (event) {
-        log.debug('client event received: %j', event);
-        var id;
-        var cbEvent;
-        try {
-            event = JSON.parse(event);
-        } catch (e) {
-            log.error('Build: invalid json: %s - ignoring', event);
-            return;
-        }
-        switch (event.type) {
-            case 'callback':
-                id = event.messageId;
-                cbEvent = pendingCallbackEvents[id];
-                assert.object(cbEvent, 'cbEvent with messageId ' + id);
-                delete pendingCallbackEvents[id];
-                if (event.error) {
-                    cbEvent.callback(new Error(event.error));
-                } else {
-                    cbEvent.callback(null, event.result);
-                }
-                break;
-            default:
-                log.error('Unhandled socket event - ignoring: %j', event);
-                break;
+
+    commitEventStream.on('readable', function commitImageOnReadable() {
+        var line;
+
+        while ((line = commitEventStream.read()) != null) {
+            log.debug('client event received: %j', line);
+            var id;
+            var cbEvent;
+            var event;
+            try {
+                event = JSON.parse(line);
+            } catch (e) {
+                log.error('Build: invalid json: %s - ignoring', line);
+                continue;
+            }
+            switch (event.type) {
+                case 'callback':
+                    id = event.messageId;
+                    cbEvent = pendingCallbackEvents[id];
+                    assert.object(cbEvent, 'cbEvent with messageId ' + id);
+                    delete pendingCallbackEvents[id];
+                    if (event.error) {
+                        cbEvent.callback(new Error(event.error));
+                    } else {
+                        cbEvent.callback(null, event.result);
+                    }
+                    break;
+                default:
+                    log.error('Unhandled socket event - ignoring: %j', event);
+                    break;
+            }
         }
     });
 
diff --git a/lib/docker-stdio.js b/lib/docker-stdio.js
index a882d6f..784116b 100644
--- a/lib/docker-stdio.js
+++ b/lib/docker-stdio.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2015, Joyent, Inc.
+ * Copyright (c) 2016, Joyent, Inc.
  */
 
 /*
@@ -30,7 +30,7 @@ var spawn = child_process.spawn;
 var execFile = child_process.execFile;
 var vmadm = require('vmadm');
 
-var LineStream = require('./linestream');
+var LineStream = require('lstream');
 var wait_flag = require('./update-wait-flag');
 
 var commands = {};
@@ -780,33 +780,38 @@ function _createLinestreamParser(opts, stream) {
         opts.log.error({ err: err }, 'LineStream threw an error');
     });
 
-    lstream.on('line', function (line) {
-        line = line.trim();
-        if (!line) {
-            return;
-        }
+    lstream.on('readable', function createLineStreamParserOnReadable() {
+        var line;
+
+        while ((line = lstream.read()) != null) {
+            line = line.trim();
+            if (!line) {
+                continue;
+            }
 
-        var parsed = JSON.parse(line);
+            var parsed = JSON.parse(line);
 
-        if (parsed.type === 'tty') {
-            if (parsed.resize) {
-                stream.resize(parsed.resize.w, parsed.resize.h);
-            } else if (parsed.data) {
+            if (parsed.type === 'tty') {
+                if (parsed.resize) {
+                    stream.resize(parsed.resize.w, parsed.resize.h);
+                } else if (parsed.data) {
+                    handleDataEvent(parsed.data);
+                }
+            } else if (parsed.type === 'stdin') {
                 handleDataEvent(parsed.data);
-            }
-        } else if (parsed.type === 'stdin') {
-            handleDataEvent(parsed.data);
-        } else if (parsed.type === 'end') {
-            tryEnd(stream);
-        } else {
-            opts.log.error({
-                parsed: parsed
-            }, 'LineStream received unrecognized data');
-            if (stream.kill) {
-                stream.kill();
-            } else {
+            } else if (parsed.type === 'end') {
                 tryEnd(stream);
+            } else {
+                opts.log.error({
+                    parsed: parsed
+                }, 'LineStream received unrecognized data');
+                if (stream.kill) {
+                    stream.kill();
+                } else {
+                    tryEnd(stream);
+                }
             }
+
         }
     });
 
@@ -1101,21 +1106,25 @@ function runContainerLogsCommand(container, params, socket) {
         log.error('LineStream threw an error %s', err.toString());
     });
 
-    lstream.on('line', function (line) {
-        line = line.trim();
-        if (!line) {
-            return;
-        }
+    lstream.on('readable', function runContainerLogsCommandOnReadable() {
+        var line;
 
-        var rec = JSON.parse(line);
-        var data;
-        if (params.Timestamps) {
-            data = rec.time + ' ' + rec.log;
-        } else {
-            data = rec.log;
-        }
+        while ((line = lstream.read()) != null) {
+            line = line.trim();
+            if (!line) {
+                continue;
+            }
 
-        writeData(rec.stream, socket, data.toString());
+            var rec = JSON.parse(line);
+            var data;
+            if (params.Timestamps) {
+                data = rec.time + ' ' + rec.log;
+            } else {
+                data = rec.log;
+            }
+
+            writeData(rec.stream, socket, data.toString());
+        }
     });
 
     cmdSpawn.on('exit', function (code, signal) {
diff --git a/lib/linestream.js b/lib/linestream.js
deleted file mode 100644
index 01a3ee8..0000000
--- a/lib/linestream.js
+++ /dev/null
@@ -1,72 +0,0 @@
-// Copyright 2013 Timothy J Fontaine <tjfontaine@gmail.com>
-//
-// Permission is hereby granted, free of charge, to any person obtaining a copy
-// of this software and associated documentation files (the 'Software'), to deal
-// in the Software without restriction, including without limitation the rights
-// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
-// copies of the Software, and to permit persons to whom the Software is
-// furnished to do so, subject to the following conditions:
-//
-// The above copyright notice and this permission notice shall be included in
-// all copies or substantial portions of the Software.
-//
-// THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
-// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
-// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
-// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
-// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
-// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
-// THE SOFTWARE
-
-/**
- * cn-agent note: Copied from
- * <https://github.com/tjfontaine/node-lstream/blob/master/linestream.js>
- * manually to not get the readable-stream dep needed only for node 0.8.
- */
-
-/*jsl:ignore*/
-/* BEGIN JSSTYLED */
-var stream = require('stream');
-var util = require('util');
-
-if (!stream.Transform)
-  stream = require('readable-stream');
-
-var LineStream = module.exports = function LineStream(opts) {
-  if (!(this instanceof LineStream)) return new LineStream(opts);
-  opts = opts || {};
-  opts.objectMode = true;
-  stream.Transform.call(this, opts);
-  this._buff = '';
-};
-util.inherits(LineStream, stream.Transform);
-
-
-LineStream.prototype._transform = function(chunk, encoding, done) {
-  var data = this._buff + chunk.toString('utf8');
-  var lines = data.split(/\r?\n|\r(?!\n)/);
-
-  this._buff = lines.pop();
-
-  var self = this;
-
-  lines.forEach(function (line) {
-    self._line(line);
-  });
-
-  done();
-};
-
-
-LineStream.prototype._flush = function(done) {
-  if (this._buff) this._line(this._buff);
-  done();
-};
-
-
-LineStream.prototype._line = function(line) {
-  this.push(line);
-  this.emit('line', line);
-};
-/* END JSSTYLED */
-/*jsl:end*/
\ No newline at end of file
diff --git a/package.json b/package.json
index 5665617..a762eab 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "cn-agent",
   "description": "Triton Compute Node Agent",
-  "version": "1.5.3",
+  "version": "1.5.4",
   "author": "Joyent (joyent.com)",
   "private": true,
   "dependencies": {
@@ -16,6 +16,7 @@
     "lazy-property": "1.0.0",
     "libuuid": "0.2.1",
     "lockfd": "1.2.0",
+    "lstream": "0.0.4",
     "mkdirp": "^0.3.4",
     "nodeunit": "0.9.1",
     "once": "1.3.0",
